python/ql/lib/semmle/python/frameworks/SqlAlchemy.qll

```diff
@@ -113,15 +113,25 @@ module SqlAlchemy {
      */
     abstract class InstanceSource extends DataFlow::LocalSourceNode { }
 
+    /**
+     * join-ordering helper for ConnectionConstruction char-pred -- without this would
+     * start with _all_ `CallCfgNode` and join those with `MethodCallNode` .. which is
+     * silly
+     */
+    pragma[noinline]
+    private DataFlow::MethodCallNode connectionConstruction_helper() {
+      result.calls(Engine::instance(), ["begin", "connect"])
+      or
+      result.calls(instance(), ["connect", "execution_options"])
+    }
+
     private class ConnectionConstruction extends InstanceSource, DataFlow::CallCfgNode {
       ConnectionConstruction() {
-        this = classRef().getACall()
-        or
-        this.(DataFlow::MethodCallNode).calls(Engine::instance(), ["begin", "connect"])
+        // without the `pragma[only_bind_out]` we would start with joining
+        // `API::Node.getACall` with `CallCfgNode` which is not optimal
+        this = pragma[only_bind_out](classRef().getACall())
```

Comment:
This comment might be true today, but there's no guarantee that it's true tomorrow, as it doesn't actually prevent an initial join between `getACall` and `CallCfgNode`.

Reply:
OTOH, this might just do the trick, as this forces `classRef` to be the initial relation in the pipeline.
```suggestion
        // without the `pragma[only_bind_out]` we would start with joining
        // `API::Node.getACall` with `CallCfgNode` which is not optimal
        this = pragma[only_bind_out](classRef()).getACall()
```

Reply:
Thanks :+1: